File: python/examples/ephemeris_comparison.py
"""JPL low-precision and VSOP2013 ephemeris comparison.

Units: MJD2000 days, metres, and metres/second.
Expected: two finite states; their frames intentionally differ.
Runtime: first VSOP evaluation is normally below 1 ms with a release wheel.
Features: default wheel with embedded VSOP2013 coefficients.
"""

import math

import numpy as np

import pykep_rust as pk


jpl = pk.Planet.jpl_low_precision("earth")
vsop = pk.Planet.vsop2013("earth_moon")
jpl_state = jpl.state(0.5)
vsop_state = vsop.state(0.5)
jpl_distance = float(np.linalg.norm(jpl_state[:3]))
vsop_distance = float(np.linalg.norm(vsop_state[:3]))
assert math.isfinite(jpl_distance)
assert math.isfinite(vsop_distance)
print(f"JPL distance [m]: {jpl_distance:.3f}")
print(f"VSOP2013 distance [m]: {vsop_distance:.3f}")
print("JPL is J2000 ecliptic; VSOP2013 is ICRF.")

File: examples/src/bin/ephemeris-comparison.rs
// Copyright (c) 2026 pykep-rust contributors
// SPDX-License-Identifier: MPL-2.0

//! Compare approximate and analytical heliocentric distance at J2000.

use pykep_core::ephemeris::{Ephemeris, JplLowPrecision, Vsop2013};
use pykep_core::math::linalg::norm;

fn main() -> pykep_core::Result<()> {
    let jpl = JplLowPrecision::new("earth")?;
    let vsop = Vsop2013::new("earth_moon")?;
    let jpl_state = jpl.state(0.5)?;
    let vsop_state = vsop.state(0.5)?;
    let jpl_distance = norm(&[jpl_state[0], jpl_state[1], jpl_state[2]])?;
    let vsop_distance = norm(&[vsop_state[0], vsop_state[1], vsop_state[2]])?;
    println!("MJD2000 0.5 heliocentric distance [m]");
    println!("JPL low precision: {jpl_distance:.3}");
    println!("VSOP2013:         {vsop_distance:.3}");
    println!("The JPL state is J2000 ecliptic; the VSOP state is ICRF.");
    Ok(())
}
